data/domain/player.tsx

```diff
@@ -703,6 +783,194 @@ export const getActivePlayerIndexDefaultFirst = (players: Player[]): number => {
     return activeIndex;
 }
 
+export const updatePlayerTalentPoints = (data: Map<string, any>) => {
+    const players = data.get("players") as Player[];
+    const account = data.get("account") as Account;
+    const alchemy = data.get("alchemy") as Alchemy;
+    const guild = data.get("guild") as Guild;
+    const bribes = data.get("bribes") as Bribe[];
+    const islandExpeditions = data.get("islandExpeditions") as IslandExpeditions;
+    const family = data.get("family") as Family;
+    const stamps = data.get("stamps") as Stamp[][];
+    const cards = data.get("cards") as Card[];
+    const taskBoard = data.get("taskboard") as TaskBoard;
+    const achievements = data.get("achievements") as Achievement[];
+    const optionListAccount = data.get("OptLacc") as number[];
+    const sigils = data.get("sigils") as Sigils;
+    const dungeonsData = data.get("dungeons") as Dungeons;
+    const breeding = data.get("breeding") as Breeding;
+    const arcade = data.get("arcade") as Arcade;
+    
+    players.forEach(player => {
+        player.talentPoints = [];
+        var talentDL: number[] = [];
+        var talentDLbonus: number[] = [];
+        var talentDN4 = -3
+
+        // Base talent points
+        for (var f = 0; 9 > f; f++) {
+            talentDN4 += player.skills.get(f+1)?.level ?? 0;
+        }
+        talentDN4 += Math.round(player.getTalentBonus(275));
+        talentDL.push(talentDN4);
+
+        for (f = talentDN4 = 0; 9 > f; f++) {
+            talentDN4 += Math.floor((player.skills.get(f+1)?.level ?? 0) / 2);
+        }
+        talentDL.push(talentDN4);
+
+        for (f = talentDN4 = 0; 9 > f; f++) {
+            talentDN4 += Math.floor((player.skills.get(f+1)?.level ?? 0) / 5);
+        }
+        talentDL.push(talentDN4);
+
+        // Bonus talent points
+        talentDLbonus.push(account.talentPointsOwned[0] ?? 0);
+        talentDLbonus.push(account.talentPointsOwned[1] ?? 0);
+        talentDLbonus.push(account.talentPointsOwned[2] ?? 0);
+        talentDLbonus.push(account.talentPointsOwned[3] ?? 0);
+        talentDLbonus.push(account.talentPointsOwned[4] ?? 0);
+        talentDLbonus.push(account.talentPointsOwned[5] ?? 0);
+
+        const bubbleArcherIndex = alchemy.cauldrons[CauldronIndex.Quicc].bubbles.findIndex(bubble => bubble.data.bonusKey == "TalArchers");
+        const bubbleWarriorIndex = alchemy.cauldrons[CauldronIndex.Power].bubbles.findIndex(bubble => bubble.data.bonusKey == "TalWarrior");
+        const bubbleMageIndex = alchemy.cauldrons[CauldronIndex.HighIQ].bubbles.findIndex(bubble => bubble.data.bonusKey == "TalWiz");
+
+        for (f = 0; 4 > f; f++) {
```

Comment:
I don't know, that's part of how Lava calculates it

The `alchemy.getBonusForPlayer(player, <bonusIndex>, <bubbleClassIndex>);` value is added to the first 4 indexes of the TalentDLBonus array, which already contains some data that comes from game data in the lines before when I retrieve bubbles indexes

If you believe it could be written more clearly don't hesitate to tell me, but this is part of the calculation as it is in-game so we'll have to keep this kind of thing

Reply:
I tried to simplify it a bit, tell me if you think something more could be done